### DISCUSSION TITLE
what is the best practice regarding useSelector

### DISCUSSION THREAD

**Discussion by send2moran** [statelyai/xstate#3886] [Q&A]
can i do something like that ?

```
const events = machineContext.useSelector((_state) => {
   return {open: _state?.context?.data.open, closed: _state?.context?.data.closed}
});


```

or to trigger a sideEffect in the selector function, is _state the latest value ?

```
const events = machineContext.useSelector(
    (_state) => {
        doSomethingElse(_state);
        return _state?.context?.data;
    },
);
```

**Comment by davidkpiano** [comment]
You should not trigger side-effects in `useSelector`. To synchronize an effect, you should use `useEffect`:

```ts
const something = machineContext.useSelector(...);

useEffect(() => {
  // some subscription
}, [something]);
```

For fire-and-forget effects, you should use XState actions.

**Reply by johnlombardo-dev** [reply]
Ever feel like people have been burned so many times by useEffect that they start to avoid useEffect when it's the right tool for the job? I think if it had been named `useSideEffect` from the start it would have made a lot more sense to everyone that assumed it was short for `useAllTheEffectsAlwaysEveryTime`.

**Reply by Andarist** [reply]
The name definitely confused many and doesn't make it easy to teach. However, I feel like `useSideEffect` would suffer from similar problems. Remember that there is no guarantee that your "side effect" will actually get executed if you manage to transition away from the given "state". In David's example, there is always a chance that `something` changes back or changes to another value and React will happily batch this into a single effect. So you can only be sure that the side effect gets executed for the latest value - but not that it gets executed for each changed, intermediate, value. That's why we call it "synchronization" - it synchronizes at some point in the future but that doesn't happen asap so there is always a chance that you might "miss" some intermediate values and that's not always desirable.

**Reply by send2moran** [reply]
Hey, Thanks for the answer, I saw patterns like that all over youtube, and I understand that no side-effects are welcome in the selector.
my question is, besides performance, 
when I return composed state, how does it work behind the scenes? 

```
const events = machineContext.useSelector((_state) => {
   return {open: _state?.context?.data.open, closed: _state?.context?.data.closed}
});


```

**Reply by davidkpiano** [reply]
@send2moran That may not work as you expect since it will return a new object reference each time.

You can use [shallow comparison](https://xstate.js.org/docs/packages/xstate-react/#shallow-comparison) to prevent unnecessary rerenders.
